libyul/backends/wasm/EWasmToText.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Component that transforms interval EWasm representation to text.
+ */
+
+#include <libyul/backends/wasm/EWasmToText.h>
+
+#include <libdevcore/StringUtils.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/range/adaptor/transformed.hpp>
+
+using namespace std;
+using namespace yul;
+
+string EWasmToText::run(
+	vector<wasm::GlobalVariableDeclaration> const& _globals,
+	vector<wasm::FunctionDefinition> const& _functions
+)
+{
+	string ret = "(module\n\n";
+	for (auto const& g: _globals)
+		ret += "    (global $" + g.variableName + " (mut i64) (i64.const 0))\n";
+	for (auto const& f: _functions)
+		ret += transform(f) + "\n";
+	return move(ret) + ")\n";
+}
+
+string EWasmToText::operator()(wasm::Literal const& _literal)
+{
+	return "(i64.const " + to_string(_literal.value) + ")";
+}
+
+string EWasmToText::operator()(wasm::LocalVariable const& _identifier)
+{
+	return "(get_local $" + _identifier.name + ")";
+}
+
+string EWasmToText::operator()(wasm::GlobalVariable const& _identifier)
+{
+	return "(get_global $" + _identifier.name + ")";
+}
+
+string EWasmToText::operator()(wasm::Label const& _label)
+{
+	return "$" + _label.name;
+}
+
+string EWasmToText::operator()(wasm::BuiltinCall const& _builtinCall)
+{
+	return "(" + _builtinCall.functionName + " " + joinTransformed(_builtinCall.arguments) + ")";
+}
+
+string EWasmToText::operator()(wasm::FunctionCall const& _functionCall)
+{
+	return "(call $" + _functionCall.functionName + " " + joinTransformed(_functionCall.arguments) + ")";
+}
+
+string EWasmToText::operator()(wasm::LocalAssignment const& _assignment)
+{
+	return "(set_local $" + _assignment.variableName + " " + visit(*_assignment.value) + ")\n";
+}
+
+string EWasmToText::operator()(wasm::GlobalAssignment const& _assignment)
+{
+	return "(set_global $" + _assignment.variableName + " " + visit(*_assignment.value) + ")\n";
+}
+
+string EWasmToText::operator()(wasm::If const& _if)
+{
+	string text = "(if " + visit(*_if.condition) + " (then\n" + indented(joinTransformed(_if.statements)) + ")";
+	if (_if.elseStatements)
+		text += "(else\n" + indented(joinTransformed(*_if.elseStatements)) + ")";
+	return std::move(text) + ")\n";
```

Comment:
Unnecessary `std::move`? (At least you did not use `std::move` in other functions's return statement for `std::string`)

Reply:
It is not a direct `return` it first adds `")\n"` at the end.

Reply:
Then should add `std::move()` in `EWasmToText::operator()(wasm::Loop const& _loop)` and `EWasmToText::operator()(wasm::Block const& _block)`'s return statement? (I just trying to point out what I observed. This issue is not important at all )

Reply:
Yes, you are right!